ledger/eval/eval.go

```diff
@@ -1444,22 +1588,63 @@ func (eval *BlockEvaluator) validateExpiredOnlineAccounts() error {
 			return fmt.Errorf("endOfBlock was unable to retrieve account %v : %w", accountAddr, err)
 		}
 
-		// true if the account is online
-		isOnline := acctData.Status == basics.Online
-		// true if the accounts last valid round has passed
-		pastCurrentRound := acctData.VoteLastValid < currentRound
-
-		if !isOnline {
-			return fmt.Errorf("endOfBlock found %v was not online but %v", accountAddr, acctData.Status)
+		if acctData.VoteID.IsEmpty() {
+			return fmt.Errorf("endOfBlock found expiration candidate %v had no vote key", accountAddr)
 		}
 
-		if !pastCurrentRound {
+		if acctData.VoteLastValid >= currentRound {
 			return fmt.Errorf("endOfBlock found %v round (%d) was not less than current round (%d)", accountAddr, acctData.VoteLastValid, currentRound)
 		}
 	}
 	return nil
 }
 
+// validateAbsentOnlineAccounts tests the accounts specified in
+// AbsentParticipationAccounts, and verifies that they need to be reset.
+func (eval *BlockEvaluator) validateAbsentOnlineAccounts() error {
```

Comment:
I don't think this has it's own test yet.

Reply:
I will have to think on this.  It will certainly take some effort to test it.

Maybe I can add a general mechanism in `DoubleLedger` that let's you manipulate a block after it's prepared, then confirm that the validator in the DoubleLedger fails in an expected way.  That could be useful for a lot of these sorts of things.

Reply:
added in `eval_test.go`
https://github.com/jannotti/go-algorand/blob/absenteeism/ledger/eval/eval_test.go#L1367